File: kh1_src/tabs/journal_tabs/chronicles.py
from dash import Dash, html, dcc, callback, Input, Output, State, ALL
import kh1_src.kh1_utils as utils


def create_chronicles():
    kh1 = utils.kh1
    chronicles = [
        html.Div([
            dcc.Markdown(list(kh1.chronicles_dict.keys())[i]),
            dcc.Checklist(
                options=[
                    {"label": f"Part {j+1}", "value": (1 << 7 - j)}\
                    for j in range(list(kh1.chronicles_dict.values())[i])
                ],
                value=[kh1.chronicles[i] & (1 << j) for j in range(8)],
                id={"type": "Chronicle", "index": i},
            )
        ]) for i in range(len(kh1.chronicles))
    ]
    return html.Div(chronicles, style={"margin-top": 20})

@callback(
    Input({"type": "Chronicle", "index": ALL}, "value"),
)
def chronicles_callback(values):
    kh1 = utils.kh1
    for i in range(len(kh1.chronicles)):
        kh1.chronicles[i] = sum(values[i])

File: kh1_src/tabs/journal_tabs/chronicles.js
export function create_chronicles() {
    const kh1jdiv = document.getElementById("kh1jdiv");
    let chronicles = `<div id="chronicles">`;
    Object.entries(window.kh1.chronicles_dict).forEach(([label, value], idx) => {
        chronicles += `
        <h3>${label}</h3>
        <label style="display: flex; align-items: center">
            <input
                type="radio"
                name=${idx}
                value=0
                ${window.kh1.chronicles[idx] == 0 ? "checked" : ""}
            >
            Locked
        </label>`;
        for (let i = 0; i < value; i++) {
            chronicles += `
            <label style="display: flex; align-items: center">
                <input
                    type="radio"
                    name=${idx}
                    value=${1 << (7 - i)}
                    ${window.kh1.chronicles[idx] & (1 << (7 - i)) ? "checked" : ""}
                >
                ${"Part " + String(i + 1)}
            </label>`;
        }
    });
    chronicles += "</div>";
    kh1jdiv.innerHTML = `
    <div>
        ${chronicles}
    </div>`;
    chronicles_callbacks();
}

function chronicles_callbacks() {
    const chronicles = document.getElementById("chronicles");
    chronicles.addEventListener("change", (e) => {
        window.kh1.chronicles[e.target.name] = e.target.value;
    });
}
